### PR TITLE
ENT-213 Consent Declined message on Offers page

### PR DESCRIPTION
**Pull request by bdero**
This change adds:

- A generic info message that can be used by passing an `info` dict to the context of any templates extending `edx/base.html`.
- A `failure_url` callback as a parameter to the consent URL.
- A consent failure message informing the user that they weren't enrolled in the course. It shows up when `consent_failure={SKU}` is passed to the coupon offer view.

**JIRA tickets**: Implements [ENT-213](https://openedx.atlassian.net/browse/ENT-213).

**Dependencies**:
This PR depends on this `edx-enterprise` PR: https://github.com/edx/edx-enterprise/pull/66

This change will come into effect with an Enterprise PR that @haikuginger will be making for ENT-185/ENT-212, where he will be including the `failure_url` call. Everything in this PR can be tested without it, though.

**Screenshots**:

*Desktop:*
desktop screenshot

*Mobile:*
mobile screenshot

**Sandbox URL**: TBD.

**Testing instructions**:

Initial setup:
1. Setup a devstack.
1. Checkout this branch (open-craft:bdero/ent-213) for ecommerce.
1. Override the edx-enterprise version with the [correct branch](https://github.com/edx/edx-enterprise/pull/66):
   ```
   /edx/bin/pip.edxapp install -e git+https://github.com/open-craft/edx-enterprise.git@bdero/add-contact-email#egg=edx-enterprise==add-contact-email
   /edx/bin/manage.edxapp lms migrate --settings=devstack
   ```
1. Connect [ecommerce to the LMS](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/ecommerce/install_ecommerce.html#configure-edx-openid-connect-oidc).
1. Setup the [course discovery](https://open-edx-course-catalog.readthedocs.io/en/latest/getting_started.html) service and [connect it to the LMS](https://open-edx-course-catalog.readthedocs.io/en/latest/oidc.html).
1. Create an Enterprise Customer in the LMS: http://localhost:8000/admin/enterprise/enterprisecustomer/add/
    1. Give it any name.
    1. Check the **Enable data sharing consent** box.
    1. For **Enforce data sharing consent**, select **At Enrollment**.
    1. Leave the **Contact email** field blank for now.
1. Add a new Course in Ecommerce: http://localhost:8002/courses/new/
    1. For the course ID, use "course-v1:edX+DemoX+Demo_Course".
    1. Name it "edX Demonstration Course".
    1. Choose "Professional Education" for the Course Type.
1. Create a new Coupon in Ecommerce: http://localhost:8002/coupons/new
    1. Give it any Coupon Name.
    1. Under Code Type choose "Enrollment Code".
    1. Give it a valid date range that includes today.
    1. For Client, type any name. E.g. "Some Client".
    1. Select the "Single course" radio button.
    1. For the Course ID, use "course-v1:edX+DemoX+Demo_Course"
    1. The only available Seat Type should be automatically populated: "Professional"
    1. For the Enterprise Customer, select the name of the customer created in step 6.

Testing the PR:
1. Navigate to the [coupon list](http://localhost:8002/coupons/) and click "Download Coupon Report" to retrieve a CSV containing the coupon URL.
    coupon report button screenshot
1. The coupon report CSV should contain an offer URL that looks like this:
    ```
    http://localhost:8002/coupons/offer/?code=EGGZYLXDKNZK6W7C
    ```
    Navigate to that URL.
1. Click on the "Enroll Now" button.
    offer landing page screenshot
    You should be redirected to the course-specific Data Sharing Consent screen.
    data sharing consent screenshot
1. Inspect the URL of the page to find that there's a `failure_url` parameter that looks something like this: `&failure_url=http%3A%2F%2Flocalhost%3A8002%2Fcoupons%2Foffer%2F%3Fcode%3DHV7SMPXDYJFTGRU2%26consent_failed%3DBB3ECA9`
1. URL decode the parameter.
    For example:
    ```
    python2 -c 'import urllib; print urllib.unquote_plus("http%3A%2F%2Flocalhost%3A8002%2Fcoupons%2Foffer%2F%3Fcode%3DHV7SMPXDYJFTGRU2%26consent_failed%3DBB3ECA9")'
    ```
    Which yields: http://localhost:8002/coupons/offer/?code=HV7SMPXDYJFTGRU2&consent_failed=BB3ECA9
1. Navigate to your decoded URL.
1. You should arrive back on the offers page. Observe that there is a blue info message area that looks similar to this:
    info message screenshot
1. Navigate back to the Enterprise Customer list in the LMS and edit the Enterprise Customer you created during the initial setup: http://localhost:8000/admin/enterprise/enterprisecustomer/
    1. Fill in the **Contact email** field with "contact@example.com".
1. Repeat steps 3 through 6.
1. Observe that the content failure message now contains the contact email:
   screenshot with contact email

**Author notes and concerns**:

The image attached in the [original ticket](https://openedx.atlassian.net/browse/ENT-213) wants to show "custom contact info". We're not currently storing this information anywhere AFAICT, so I've excluded it from this PR for the time being. CC @bradenmacdonald @haikuginger 

**Reviewers**
- [ ] @pomegranited 
- [ ] edX reviewer[s] TBD

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @bdero! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fecommerce&number=1189"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

File: ecommerce/coupons/views.py [hunk 1]
```diff
@@ -109,6 +111,38 @@ def get_context_data(self, **kwargs):
                 return {'error': _('The voucher is not applicable to your current basket.')}
             valid_voucher, msg = voucher_is_valid(voucher, products, self.request)
             if valid_voucher:
+                consent_failed = self.request.GET.get('consent_failed', None)
```
Comment (pomegranited) [inline ecommerce/coupons/views.py]: `'consent_failed'` is an important magic string.  Can you set a variable somewhere that this function (and `get_enterprise_course_consent_url`) can share?

Up for discussion.

File: ecommerce/static/js/routers/offer_router.js [hunk 2]
```diff
@@ -11,11 +11,15 @@ define([
...
-                'offer/?code=:code': 'showOfferPage'
+            initialize: function(options) {
+                // This is where views will be rendered
+                this.$el = options.$el;
+
+                this.route(/^offer\/\?code=(\w+)/, 'showOfferPage');
             },
 
             showOfferPage: function(code) {
+                console.log(code);
```
Comment (pomegranited) [inline ecommerce/static/js/routers/offer_router.js]: nit: remove logging.

File: ecommerce/coupons/views.py [hunk 3]
```diff
@@ -109,6 +111,38 @@ def get_context_data(self, **kwargs):
...
+                if consent_failed is not None:
+                    # The user is redirected to this view with `consent_failed` defined (as the product SKU) when the
+                    # user doesn't consent to data sharing.
+                    try:
+                        product = StockRecord.objects.get(partner_sku=consent_failed).product
+                    except StockRecord.DoesNotExist:
+                        return {'error': _('SKU {sku} does not exist.').format(sku=consent_failed)}
+
+                    self.template_name = 'coupons/offer.html'
```
Comment (pomegranited) [inline ecommerce/coupons/views.py]: Line duplicated from original [L146](https://github.com/edx/ecommerce/pull/1189/files#diff-c797ef6707ae3dfb1ee8828592244a9eR146).

File: ecommerce/coupons/views.py (hunk 1 above)
Comment (pomegranited) [inline ecommerce/coupons/views.py]: I know we try to tread lightly through the code we modify, but this method is getting very hard to parse.  Would you consider a slight rewrite, to keep the error messages towards the top?  Also removes the duplicate line noted below.

```python
    def get_context_data(self, **kwargs):
        code = self.request.GET.get('code', None)
        if code is None:
            return {'error': _('This coupon code is invalid.')}

        try:
            voucher, products = get_voucher_and_products_from_code(code=code)
        except Voucher.DoesNotExist:
            return {'error': _('Coupon does not exist')}
        except exceptions.ProductNotFoundError:
            return {'error': _('The voucher is not applicable to your current basket.')}
        valid_voucher, msg = voucher_is_valid(voucher, products, self.request)
        if not valid_voucher:
            return {'error': msg}

        context_data = None
        consent_failed = self.request.GET.get('consent_failed', None)
        if consent_failed is not None:
            # The user is redirected to this view with `consent_failed` defined (as the product SKU) when the
            # user doesn't consent to data sharing.
            try:
                product = StockRecord.objects.get(partner_sku=consent_failed).product
            except StockRecord.DoesNotExist:
                return {'error': _('SKU {sku} does not exist.').format(sku=consent_failed)}

            # Return the view with an info message informing the user that the enrollment didn't complete
            request = get_current_request()
            enterprise_customer = get_enterprise_customer_from_voucher(
                request.site,
                request.user.access_token,
                voucher
            )
            context_data = {
                'info': {
                    'title': _('Enrollment in {course_name} was not complete.').format(
                        course_name=product.course.name
                    ),
                    'message': _(
                        'If you have concerns about sharing your data, please contact your administrator '
                        'at {enterprise}.'
                    ).format(
                        enterprise=enterprise_customer['name']
                    )
                }
            }

        self.template_name = 'coupons/offer.html'
        return context_data
```

**Review by pomegranited — CHANGES_REQUESTED** [review]
Works as described!  Thank you for the detailed test instructions.

Once my comments are addressed or refuted, I'm happy to approve.

- [x] I tested this on my devstack, using Chrome and the mobile emulator.
- [x] I read through the code, and left some comments.
- [x] I checked for accessibility issues.  The added info bar is a static part of the HTML template, and so screen readers should have no trouble reading it.
- [x] Includes documentation - docstrings.

File: ecommerce/coupons/views.py (hunk 1 above)
Comment (bdero) [reply ecommerce/coupons/views.py]: Alrighty, I've renamed the variable to `consent_failed_sku`, since that's all it is at the end of the day. The reason we need to SKU to be passed for this info message is that we need a way to identify which product (which has the course info) the user denied consent for in the list of offers.

I've kept the name `consent_failed` for the GET param though, since I think it makes sense to say that consent gathering failed for a specified SKU. Does this seem sound?

Let me know if this addresses your concern!

Comment (bdero) [reply ecommerce/coupons/views.py]: Totally agree, I'm all for this and added these changes in. Thanks!

File: ecommerce/coupons/views.py (hunk 3 above)
Comment (bdero) [reply ecommerce/coupons/views.py]: All set, thanks!

File: ecommerce/static/js/routers/offer_router.js (hunk 2 above)
Comment (bdero) [reply ecommerce/static/js/routers/offer_router.js]: Removed, thanks!

**Comment by bdero** [comment]
@pomegranited Thank you for the review! I've addressed all comments.

File: ecommerce/coupons/views.py (hunk 1 above)
Comment (bdero) [reply ecommerce/coupons/views.py]: Ah, nevermind, I just got what you were saying here. I added `CONSENT_FAILED_PARAM` to `ecommerce.enterprise.utils`, which is where `get_enterprise_course_consent_url` lives.

Comment (pomegranited) [reply ecommerce/coupons/views.py]: +1 Yep that's what I meant.  Sorry for being unclear!

**Review by pomegranited — APPROVED** [review]

**Comment by bdero** [comment]
FYI @mattdrayer @asadiqbal08 @saleem-latif, this PR along with https://github.com/edx/edx-enterprise/pull/66 and is ready for your review!

**Comment by mattdrayer** [comment]
LGTM @bdero!  Definitely would like @asadiqbal08 to take a look here -- also we should see if @vkaracic or @marjev can take a swing through as well since this is an ecommerce mod.

File: ecommerce/enterprise/utils.py [hunk 4]
```diff
@@ -54,6 +55,7 @@ def get_enterprise_customer(site, token, uuid):
         'name': response['name'],
         'id': response['uuid'],
         'enable_data_sharing_consent': response['enable_data_sharing_consent'],
+        'contact_email': response['contact_email'],
```
Comment (pomegranited) [inline ecommerce/enterprise/utils.py]: I'd change this to reduce the dependency on https://github.com/edx/edx-enterprise/pull/66:

```python
'contact_email': response.get('contact_email', ''),
```

File: ecommerce/coupons/views.py [hunk 5]
```diff
@@ -100,20 +103,58 @@ class CouponOfferView(TemplateView):
...
+        except Voucher.DoesNotExist:
+            return {'error': _('Coupon does not exist')}
+        except exceptions.ProductNotFoundError:
+            return {'error': _('The voucher is not applicable to your current basket.')}
+        valid_voucher, msg = voucher_is_valid(voucher, products, self.request)
+        if not valid_voucher:
             return {'error': msg}
-        return {'error': _('This coupon code is invalid.')}
+
+        context_data = None
```
Comment (pomegranited) [inline ecommerce/coupons/views.py]: At the risk of being annoying and asking for another refactor.. I think the functionality from here on down should be removed to a new method in `ecommerce.enterprise.utils`, e.g.

```python
        context_data = get_enterprise_customer_context_data(self.request, voucher)
```

`get_enterprise_customer_context_data` can return None if there's nothing to say.

**Review by pomegranited — CHANGES_REQUESTED** [review]
I've got one small nit, and another impertinent refactor request..  🐑

File: ecommerce/enterprise/utils.py (hunk 4 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py]: Great idea! Thanks!

File: ecommerce/coupons/views.py (hunk 5 above)
Comment (bdero) [reply ecommerce/coupons/views.py]: Yeah, I think this is a good idea - this seems like it should live in the Enterprise module. I went ahead and made `get_enterprise_customer_consent_failed_context_data` in there. It's kind of a mouthful, but I wasn't sure how to make it shorter while still being true to what it does. Thanks!

**Comment by bdero** [comment]
Thanks for the review @pomegranited and thanks for taking a look @mattdrayer! I've addressed all comments.

File: ecommerce/coupons/tests/test_views.py [hunk 6]
```diff
@@ -238,17 +242,25 @@ def test_no_product(self):
...
         # Make sure to always pair `course_seat_types` and `catalog_query` parameters because
         # if one of them is missing it could result in a SEGFAULT error when running tests
         # with migrations enabled.
-        _range = RangeFactory(products=[credit_seat, ], course_seat_types='credit', catalog_query='*:*')
+        range_kwargs = {
+            'products': [credit_seat, ],
+            'course_seat_types': 'credit',
+            'catalog_query': '*:*',
+        }
+        if enterprise_customer is not None:
```
Comment (vkaracic) [inline ecommerce/coupons/tests/test_views.py]: ``if enterprise_customer:``

File: ecommerce/coupons/views.py [hunk 7]
```diff
@@ -100,20 +102,25 @@ class CouponOfferView(TemplateView):
...
+        except exceptions.ProductNotFoundError:
+            return {'error': _('The voucher is not applicable to your current basket.')}
+        valid_voucher, msg = voucher_is_valid(voucher, products, self.request)
+        if not valid_voucher:
             return {'error': msg}
-        return {'error': _('This coupon code is invalid.')}
+
+        context_data = get_enterprise_customer_consent_failed_context_data(self.request, voucher)
+
+        if context_data is not None and 'error' not in context_data.keys():
```
Comment (vkaracic) [inline ecommerce/coupons/views.py]: ``if context_data and 'error' not in context_data:``

File: ecommerce/enterprise/utils.py [hunk 8]
```diff
@@ -74,6 +78,53 @@ def get_enterprise_customers(site, token):
...
 
+def get_enterprise_customer_consent_failed_context_data(request, voucher):
+    """
+    Get the template context to display a message informing the user that they were not enrolled in the course
+    due to not consenting to data sharing with the Enterprise Customer.
+
+    If the `consent_failed` GET param is defined and it's not set to a valid SKU, return an error context that
+    says the given SKU doesn't exist.
+    """
+    consent_failed_sku = request.GET.get(CONSENT_FAILED_PARAM, None)
```
Comment (vkaracic) [inline ecommerce/enterprise/utils.py]: You don't need the additional ``None`` parameter when doing ``dict.get()``

File: ecommerce/enterprise/utils.py [hunk 9]
```diff
@@ -74,6 +78,53 @@ def get_enterprise_customers(site, token):
...
+    except StockRecord.DoesNotExist:
+        return {'error': _('SKU {sku} does not exist.').format(sku=consent_failed_sku)}
+
+    # Return the view with an info message informing the user that the enrollment didn't complete
+    enterprise_customer = get_enterprise_customer_from_voucher(
+        request.site,
+        request.user.access_token,
+        voucher
+    )
+    if len(enterprise_customer['contact_email']) > 0:
```
Comment (vkaracic) [inline ecommerce/enterprise/utils.py]: ``if enterprise_customer['contact_email']: ``

File: ecommerce/enterprise/utils.py [hunk 10]
```diff
@@ -127,21 +178,32 @@ def get_enterprise_course_consent_url(site, code, sku, consent_token, course_id,
...
         params=urlencode({
             'code': code,
             'sku': sku,
             'consent_token': consent_token,
         })
     )
+    failure_url = '{base}{resource}?code={code}&{consent_failed_param_name}={sku}'.format(
+        base=base_url,
+        resource=reverse('coupons:offer'),
+        code=code,
```
Comment (vkaracic) [inline ecommerce/enterprise/utils.py]: Please use ``urlencode`` for encoding query string parameters.

File: ecommerce/static/js/routers/offer_router.js [hunk 11]
```diff
@@ -11,8 +11,11 @@ define([
             // Base/root path of the app
             root: '/coupons/',
 
-            routes: {
-                'offer/?code=:code': 'showOfferPage'
+            initialize: function(options) {
```
Comment (vkaracic) [inline ecommerce/static/js/routers/offer_router.js:14]: Why this change?

File: ecommerce/enterprise/utils.py [hunk 12]
```diff
@@ -74,6 +78,53 @@ def get_enterprise_customers(site, token):
     ]
 
 
+def get_enterprise_customer_consent_failed_context_data(request, voucher):
```
Comment (marjev) [inline ecommerce/enterprise/utils.py:82]: Will this method ever be used outside the `CouponOfferView`? If not, this method should perhaps live within the view itself.

File: ecommerce/templates/edx/partials/_info_message.html [hunk 13]
```diff
@@ -0,0 +1,11 @@
+{% if info != None %}
```
Comment (marjev) [inline ecommerce/templates/edx/partials/_info_message.html]: I think only `if info` will do the trick.

File: ecommerce/coupons/views.py [hunk 14]
```diff
@@ -100,20 +102,25 @@ class CouponOfferView(TemplateView):
...
+        except Voucher.DoesNotExist:
+            return {'error': _('Coupon does not exist')}
+        except exceptions.ProductNotFoundError:
+            return {'error': _('The voucher is not applicable to your current basket.')}
+        valid_voucher, msg = voucher_is_valid(voucher, products, self.request)
+        if not valid_voucher:
             return {'error': msg}
-        return {'error': _('This coupon code is invalid.')}
+
+        context_data = get_enterprise_customer_consent_failed_context_data(self.request, voucher)
```
Comment (vkaracic) [inline ecommerce/coupons/views.py]: Make sure you're not overriding default context data. First get the default context data with ``context = super(CouponOfferView, self).get_context_data(**kwargs)`` and then append your data to that dictionary.

File: ecommerce/coupons/tests/test_views.py [hunk 15]
```diff
@@ -270,6 +282,40 @@ def test_credit_seat_response(self):
...
+    @ddt.unpack
+    def test_consent_failed_message(self, contact_email, expected_response):
+        """ Verify that the consent failure message shows up when the consent_failed parameter is a valid SKU. """
+        self.mock_specific_enterprise_customer_api(
+            ENTERPRISE_CUSTOMER,
+            name='TestShib',
+            contact_email=contact_email
+        )
+        base_url = self.prepare_url_for_credit_seat(enterprise_customer=ENTERPRISE_CUSTOMER)
+        product = self.credit_seat.rangeproduct_set.first().product
```
Comment (vkaracic) [inline ecommerce/coupons/tests/test_views.py]: Isn't the seat the product? You can do ``sku = self.credit_seat.stockrecords.first().partner_sku`` directly.

**Review by vkaracic — CHANGES_REQUESTED** [review]
I have a few questions.

**Review by marjev — CHANGES_REQUESTED** [review]

**Review by pomegranited — APPROVED** [review]
:+1: when tests pass.  Thanks for your patience, @bdero !

- [x] I tested this on my devstack using the lovely detailed instructions provided.
- [x] I read through the code.
- [x] I checked for accessibility issues.
- [x] Includes documentation.

File: ecommerce/coupons/tests/test_views.py (hunk 6 above)
Comment (bdero) [reply ecommerce/coupons/tests/test_views.py]: Done. I believe I've gotten comments on other edX PRs asking me to explicitly do checks against `None` instead of doing implicit truthy checks (I might not be remembering correctly, though). I will change this for now, but let me know if you still think this is the right way to go.

File: ecommerce/coupons/tests/test_views.py (hunk 15 above)
Comment (bdero) [reply ecommerce/coupons/tests/test_views.py]: It's not just grabbing the `product` for the SKU, the test also requires the course name for verification of the view message.

Given the product line wouldn't go away, would you still prefer `self.credit_seat.stockrecords.first().partner_sku` over `sku = product.stockrecords.first().partner_sku` in this case?

File: ecommerce/coupons/views.py (hunk 14 above)
Comment (bdero) [reply ecommerce/coupons/views.py]: Done, thanks!

File: ecommerce/coupons/views.py (hunk 7 above)
Comment (bdero) [reply ecommerce/coupons/views.py]: Done.

File: ecommerce/enterprise/utils.py (hunk 8 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py]: Fixed, thanks.

File: ecommerce/enterprise/utils.py (hunk 9 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py]: Done.

File: ecommerce/enterprise/utils.py (hunk 10 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py]: Done.

File: ecommerce/static/js/routers/offer_router.js (hunk 11 above)
Comment (bdero) [reply ecommerce/static/js/routers/offer_router.js:14]: This change is necessary because any URL containing more params than just `code`, e.g. `offer/?code=SOME_CODE&consent_failed=SOME_SKU`, would result in those params getting parsed as the code before being passed to the backbone view.

I think it's safe to assume that `codes` will at least never have non-alphanumeric characters, so my fix was to more accurately parse out the code param as `\w+`.

File: ecommerce/enterprise/utils.py (hunk 12 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py:82]: I highly doubt it's usable outside the context of `CouponOfferView`, but the general trend seems to be that most Enterprise stuff is pulled into the Enterprise module whether that's the case or not. What are your thoughts about this @pomegranited @mattdrayer?

File: ecommerce/templates/edx/partials/_info_message.html (hunk 13 above)
Comment (bdero) [reply ecommerce/templates/edx/partials/_info_message.html]: Done.

**Comment by bdero** [comment]
@pomegranited Thanks!
@vkaracic @marjev Thanks for the reviews, I've addressed all your comments!

File: ecommerce/enterprise/utils.py (hunk 12 above)
Comment (pomegranited) [reply ecommerce/enterprise/utils.py:82]: My thoughts were along the same lines as @bdero, but I see your point, since this logic is specific to this view. Your call, @marjev.

**Comment by mattdrayer** [comment]
@asadiqbal08 if you can do a review of this PR, I think @bdero will be ready to wrap things up.

File: ecommerce/enterprise/utils.py [hunk 16]
```diff
@@ -74,6 +79,58 @@ def get_enterprise_customers(site, token):
...
+        voucher
+    )
+    if not enterprise_customer:
+        return {'error': _('There is no Enterprise Customer associated with SKU {sku}.').format(
+            sku=consent_failed_sku
+        )}
+
+    if enterprise_customer['contact_email']:
+        contact_info = ' at {contact_email}'.format(contact_email=enterprise_customer['contact_email'])
+    else:
```
Comment (asadiqbal08) [inline ecommerce/enterprise/utils.py]: nit: We can remove the extra `else`
```
contact_info = ''
if enterprise_customer['contact_email']:
   contact_info = ' at {contact_email}'.format(contact_email=enterprise_customer['contact_email'])

```

File: ecommerce/enterprise/utils.py [hunk 17]
```diff
@@ -74,6 +79,58 @@ def get_enterprise_customers(site, token):
...
+        contact_info = ''
```
Comment (asadiqbal08) [inline ecommerce/enterprise/utils.py]: a question: If we could not find out the enterprise_customer any `contact_email` then any other default one to show instead of empty sting ?

File: ecommerce/enterprise/utils.py [hunk 18]
```diff
@@ -74,6 +79,58 @@ def get_enterprise_customers(site, token):
...
+
+    return {
+        'info': {
+            'title': _('Enrollment in {course_name} was not complete.').format(
+                course_name=product.course.name
+            ),
+            'message': _(
+                'If you have concerns about sharing your data, please contact your administrator '
+                'at {enterprise}{contact_info}.'
```
Comment (asadiqbal08) [inline ecommerce/enterprise/utils.py]: A mock-up on JIRA shows a message `If you have concerns about sharing your data, please contact your administrator at {enterprise} at {contact_info}.`

**Comment by asadiqbal08** [comment]
@bdero few questions above. rest of the changes LGTM 👍 . branch need to be rebase with latest

File: ecommerce/enterprise/utils.py (hunk 17 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py]: No, on the upstream ticket we decided for it to be an optional contact email field, so we need to gracefully handle the case when it's blank.

File: ecommerce/enterprise/utils.py (hunk 18 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py]: Hmm yea, I think this should actually be refactored to be two separate translation strings - one for each case.

The first looks like this:
> If you have concerns about sharing your data, please contact your administrator at {enterprise} at {contact_info}.

And the second looks like this:
> If you have concerns about sharing your data, please contact your administrator at {enterprise}.

The way I wrote it is incorrect because the "at" part of the string isn't going through i18n.

File: ecommerce/enterprise/utils.py (hunk 12 above)
Comment (marjev) [reply ecommerce/enterprise/utils.py:82]: IMHO, `get_enterprise_customer_from_voucher` seems to be related to the Enterprise stuff from the business logic perspective, so it should live in the Enterprise module, but since everything else in this method is about formatting data specifically for the `CouponOfferView`, I think it should be close to the View itself.
But, I'll leave it to @bdero to decide what he thinks is best in the context of this PR, to keep things consistent. It's not a showstopper.

**Review by marjev — APPROVED** [review]

File: ecommerce/enterprise/utils.py (hunk 12 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py:82]: Thanks. I don't have a terribly strong opinion here, but I'll leave this in `enterprise/utils.py` for now just because it seems nice to have e.g. the enterprise-related errors bundled up and the docstring the in this case. The way I had it embedded in the view before was starting to become quite hard to parse through. :)

File: ecommerce/enterprise/utils.py (hunk 16 above)
Comment (bdero) [reply ecommerce/enterprise/utils.py]: Done.

**Comment by bdero** [comment]
All comments addressed @asadiqbal08, thanks for the review!

@vkaracic When you get a chance, can you give this another look over and approve if you have no more requests?

File: ecommerce/coupons/tests/test_views.py (hunk 15 above)
Comment (vkaracic) [reply ecommerce/coupons/tests/test_views.py]: I'm trying to tell you that ``self.credit_seat`` is the product. In this case this should be true:
```
self.credit_seat.rangeproduct_set.first().product == self.credit_seat
```

File: ecommerce/coupons/views.py [hunk 19]
```diff
@@ -100,20 +102,26 @@ class CouponOfferView(TemplateView):
...
+        except Voucher.DoesNotExist:
+            return {'error': _('Coupon does not exist')}
+        except exceptions.ProductNotFoundError:
+            return {'error': _('The voucher is not applicable to your current basket.')}
+        valid_voucher, msg = voucher_is_valid(voucher, products, self.request)
+        if not valid_voucher:
             return {'error': msg}
-        return {'error': _('This coupon code is invalid.')}
+
+        context_data = super(CouponOfferView, self).get_context_data(**kwargs) or {}
```
Comment (vkaracic) [inline ecommerce/coupons/views.py]: No need for the ``or``, context will always be a dictionary.

File: ecommerce/coupons/views.py [hunk 20]
```diff
@@ -100,20 +102,26 @@ class CouponOfferView(TemplateView):
...
+        context_data.update(get_enterprise_customer_consent_failed_context_data(self.request, voucher))
+
+        if context_data and 'error' not in context_data.keys():
```
Comment (vkaracic) [inline ecommerce/coupons/views.py]: No need for ``.keys()``.

File: ecommerce/static/js/routers/offer_router.js (hunk 11 above)
Comment (vkaracic) [reply ecommerce/static/js/routers/offer_router.js:14]: Have you tried ``'offer/?code=:code/': 'showOfferPage'`` ?

File: ecommerce/coupons/tests/test_views.py (hunk 6 above)
Comment (vkaracic) [reply ecommerce/coupons/tests/test_views.py]: In my experience it is. Although Python's ethos is ``explicit is better than implicit``, I believe saying ``if something-exists`` is explicit enough since ``enterprise_customer`` is not a boolean value. Even if it were I doubt you would want to proceed if it's ``False``.

**Review by vkaracic — APPROVED** [review]
A couple of nitpicks left.

File: ecommerce/coupons/views.py (hunk 20 above)
Comment (bdero) [reply ecommerce/coupons/views.py]: Done.

File: ecommerce/coupons/views.py (hunk 19 above)
Comment (bdero) [reply ecommerce/coupons/views.py]: Done, thanks.

File: ecommerce/static/js/routers/offer_router.js (hunk 11 above)
Comment (bdero) [reply ecommerce/static/js/routers/offer_router.js:14]: I gave this a try, but this would only match against a URL containing a slash at the end of the code:
```
offer/?code=HV7SMPXDYJFTGRU2/
```

I don't think the routing library backbone uses is intended for grabbing stuff out of get params. I'd call this a generally bad pattern. An example of another downside of writing route keys this way is that now the order of the get params is needlessly important.

File: ecommerce/coupons/tests/test_views.py (hunk 15 above)
Comment (bdero) [reply ecommerce/coupons/tests/test_views.py]: Gotcha, fixed. Thanks!

**Comment by bdero** [comment]
Thanks @vkaracic !

**Comment by vkaracic** [comment]
You're very welcome!
